src/GameState.cpp

```diff
@@ -320,6 +320,48 @@ std::optional<std::optional<Color>> GameState::getOutcome() const {
     // TODO:
     // If we're in check, see if there's a way to get out. (If not, our opponent wins.)
     // If we're not in check, see if there's any move that doesn't put us in check. (If not, it's stalemate.)
+
+
+    // Check for checkmate if king is in check
+    if (this->isCheck(this->getTurn()))
+    {
+        int legalMoveFlag = 0;
+        for (const auto [square, piece] : pieceMap) {
+            // Iterating over the pieces of the same color and checking for legal moves
+            if (piece.color != this->turn) continue;
+
+            for (auto move : this->generateMoves(square)) {
+                if(isLegal(move))
+                    legalMoveFlag++;
```

Comment:
You can break out of the loop as soon as you find a legal move. It's also sufficient for `legalMoveFlag` to be a bool, since we don't care about the number, just whether one exists.
Going further: you can go ahead and `return std::nullopt` when you find a legal move, which eliminates the need for an extra variable altogether.